### PR TITLE
Add qubits as a param of PauliSum.matrix()

### PR DESCRIPTION
**Pull request by madcpf**

### CODE REVIEW

**Review by vtomole — DISMISSED** [review]
What is the cause  for this change? I see that it doesn't matter whether or not you pass the qubits, the resulting matrix will be the same:

```python
assert np.allclose(H1, paulisum1.matrix())
assert np.allclose(H1, paulisum1.matrix([q[0], q[1]]))
assert np.allclose(H1, paulisum1.matrix([q[1], q[0]]))
```

**Comment by madcpf** [comment]
> What is the cause for this change? I see that it doesn't matter whether or not you pass the qubits, the resulting matrix will be the same:
> 
> ```python
> assert np.allclose(H1, paulisum1.matrix())
> assert np.allclose(H1, paulisum1.matrix([q[0], q[1]]))
> assert np.allclose(H1, paulisum1.matrix([q[1], q[0]]))
> ```

Hi Victory, thanks for looking into this. It makes a difference for H2, and does not affect H1 or H3, which is as expected. The use case for this change is https://github.com/quantumlib/ReCirq/pull/174, where non-alphabetical sequence of qubits might be inputs of pauli_decomposition().

**Comment by tanujkhattar** [comment]
@madcpf Since PauliSum represents an operation, an instance of PauliSum represents a definite operation on a particular set of qubits. If you want the same operation applied to a different set of qubits, you should use `ps.with_qubits(<new_qubits_ordering>)`. 

Specifically, from the examples used in your test, 

```
ps1 = cirq.X(q[0]) * cirq.X(q[1]) + cirq.Z(q[0])
# H2 : Matrix for X(0) * X(1) + Z(0)
H2 = np.array([[1.0, 0.0, 0.0, 1.0], [0.0, 1.0, 1.0, 0.0], [0.0, 1.0, -1.0, 0.0], [1.0, 0.0, 0.0, -1.0]])
assert np.allclose(H2, ps1.matrix()) 

# H2_2 : Matrix for X(1) * X(0) + Z(1) --> this is a different operation!
H2_2 = np.array([[1.0, 0.0, 0.0, 1.0], [0.0, -1.0, 1.0, 0.0], [0.0, 1.0, 1.0, 0.0], [1.0, 0.0, 0.0, -1.0]])
# ps2 now represents X(1) * X(0) + Z(1) operation
ps2 = ps1.with_qubits(q[1], q[0]) 
assert np.allclose(H2_2, ps2.matrix())  # passes as expected. 
```

Hope this satisfies your use case. I'll close this PR for now, please feel free to open it again if you have any follow up questions / comments.

**Comment by madcpf** [comment]
> @madcpf Since PauliSum represents an operation, an instance of PauliSum represents a definite operation on a particular set of qubits. If you want the same operation applied to a different set of qubits, you should use `ps.with_qubits(<new_qubits_ordering>)`.
> 
> Specifically, from the examples used in your test,
> 
> ```python
> ps1 = cirq.X(q[0]) * cirq.X(q[1]) + cirq.Z(q[0])
> # H2 : Matrix for X(0) * X(1) + Z(0)
> H2 = np.array([[1.0, 0.0, 0.0, 1.0], [0.0, 1.0, 1.0, 0.0], [0.0, 1.0, -1.0, 0.0], [1.0, 0.0, 0.0, -1.0]])
> assert np.allclose(H2, ps1.matrix()) 
> 
> # H2_2 : Matrix for X(1) * X(0) + Z(1) --> this is a different operation!
> H2_2 = np.array([[1.0, 0.0, 0.0, 1.0], [0.0, -1.0, 1.0, 0.0], [0.0, 1.0, 1.0, 0.0], [1.0, 0.0, 0.0, -1.0]])
> # ps2 now represents X(1) * X(0) + Z(1) operation
> ps2 = ps1.with_qubits(q[1], q[0]) 
> assert np.allclose(H2_2, ps2.matrix())  # passes as expected. 
> ```
> 
> Hope this satisfies your use case. I'll close this PR for now, please feel free to open it again if you have any follow up questions / comments.

Thanks Tanuj for looking into this! Yes that will satisfy my use case. But I have problem in using it:

In [74]: ps1 = cirq.X(q[0]) * cirq.X(q[1]) + cirq.Z(q[0])
    ...: 

In [75]: ps2 = ps1.with_qubits(q[1], q[0])
    ...: 
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-75-bf0735fea75c> in <module>
----> 1 ps2 = ps1.with_qubits(q[1], q[0])

AttributeError: 'PauliSum' object has no attribute 'with_qubits'

What might be the problem here?

Thanks!

**Comment by tanujkhattar** [comment]
Maybe check your cirq version? 

Here's a colab demonstrating that the above pasted code works as expected: https://colab.research.google.com/drive/13VbwzvWXHOVgyiXjK811_E0W444H5Zib?usp=sharing

The feature was added in this PR: https://github.com/quantumlib/Cirq/pull/3427

**Comment by madcpf** [comment]
Yes that's the problem. Thanks a lot Tanuj!

**Comment by madcpf** [comment]
Hi Tanuj,

The functionality I actually need is to change cirq.X(q[0]) * cirq.X(q[1]) +
cirq.Z(q[0]) into cirq.X(q[1]) * cirq.X(q[0]) + cirq.Z(q[0]), i.e. they are
essentially the same operation, but their matrix representation is
different. Is this functionality currently supported by the framework?
AFAIK for some reason it always reorder the PauliString alphabetically...

Thanks,
Pengfei

**Comment by madcpf** [comment]
> Maybe check your cirq version?
> 
> Here's a colab demonstrating that the above pasted code works as expected on the latest cirq release: https://colab.research.google.com/drive/13VbwzvWXHOVgyiXjK811_E0W444H5Zib?usp=sharing
> 
> The feature was added in this PR: #3427

Hi Tanuj,

The functionality I actually need is to change cirq.X(q[0]) * cirq.X(q[1]) + cirq.Z(q[0]) into cirq.X(q[1]) * cirq.X(q[0]) + cirq.Z(q[0]), i.e. they are essentially the same operation, but their matrix representation is different. Is this functionality currently supported by the framework? AFAIK for some reason it always reorder the PauliString alphabetically...

Thanks,
Pengfei

**Comment by tanujkhattar** [comment]
> change `cirq.X(q[0]) * cirq.X(q[1]) + cirq.Z(q[0])` into `cirq.X(q[1]) * cirq.X(q[0]) + cirq.Z(q[0])`

I'm not sure I understand the use case. If `X(q[0]) * Y(q[1]) + Z(q[0])` is your input, then what would you want the output to be? 
- Option-1: `X(q[1]) * Y(q[0]) + Z(q[1])`. In this case, it's the same operation but applied on different ordering of qubits. This is what `ps.with_qubits(q[1], q[0])` should give you. Note that your proposed change in the `matrix` function also does this. 

- Option-2: `X(q[1]) * Y(q[0]) + Z(q[0])` : This is identical to the example you gave above where you've flipped the qubits only for the first pauli string but not on the second? This is a different pauli string altogether and it's not clear for which terms you are flipping the qubits if there are more than one terms?

**Comment by madcpf** [comment]
> > change `cirq.X(q[0]) * cirq.X(q[1]) + cirq.Z(q[0])` into `cirq.X(q[1]) * cirq.X(q[0]) + cirq.Z(q[0])`
> 
> I'm not sure I understand the use case. If `X(q[0]) * Y(q[1]) + Z(q[0])` is your input, then what would you want the output to be?
> 
> * Option-1: `X(q[1]) * Y(q[0]) + Z(q[1])`. In this case, it's the same operation but applied on different ordering of qubits. This is what `ps.with_qubits(q[1], q[0])` should give you. Note that your proposed change in the `matrix` function also does this.
> * Option-2: `X(q[1]) * Y(q[0]) + Z(q[0])` : This is identical to the example you gave above where you've flipped the qubits only for the first pauli string but not on the second? This is a different pauli string altogether and it's not clear for which terms you are flipping the qubits if there are more than one terms?

Thanks Tanuj for the quick response!

What I'm proposing is actually 

Option-3: Y(q[1]) * X(q[0]) + Z(q[0]), i.e only change the order of multiplication, since the order of matrices matters in the Kronecker product, and will give different matrices as output.

I think it's already supported in PauliStrings, but not in PauliSum yet:
In [92]: ps2 = cirq.X(q[0]) * cirq.Y(q[1])

In [93]: ps2.matrix([q[0], q[1]])
Out[93]: 
array([[0.+0.j, 0.+0.j, 0.+0.j, 0.-1.j],
       [0.+0.j, 0.+0.j, 0.+1.j, 0.+0.j],
       [0.+0.j, 0.-1.j, 0.+0.j, 0.+0.j],
       [0.+1.j, 0.+0.j, 0.+0.j, 0.+0.j]])

In [94]: ps2.matrix([q[1], q[0]])
Out[94]: 
array([[0.+0.j, 0.+0.j, 0.+0.j, 0.-1.j],
       [0.+0.j, 0.+0.j, 0.-1.j, 0.+0.j],
       [0.+0.j, 0.+1.j, 0.+0.j, 0.+0.j],
       [0.+1.j, 0.+0.j, 0.+0.j, 0.+0.j]])

My relevant PR that would like to use this functionality is https://github.com/quantumlib/ReCirq/pull/174/files.

**Comment by madcpf** [comment]
Hi Tanuj, may I have your opinion on this please? Thanks!

**Comment by madcpf** [comment]
> > change `cirq.X(q[0]) * cirq.X(q[1]) + cirq.Z(q[0])` into `cirq.X(q[1]) * cirq.X(q[0]) + cirq.Z(q[0])`
> 
> I'm not sure I understand the use case. If `X(q[0]) * Y(q[1]) + Z(q[0])` is your input, then what would you want the output to be?
> 
> * Option-1: `X(q[1]) * Y(q[0]) + Z(q[1])`. In this case, it's the same operation but applied on different ordering of qubits. This is what `ps.with_qubits(q[1], q[0])` should give you. Note that your proposed change in the `matrix` function also does this.
> * Option-2: `X(q[1]) * Y(q[0]) + Z(q[0])` : This is identical to the example you gave above where you've flipped the qubits only for the first pauli string but not on the second? This is a different pauli string altogether and it's not clear for which terms you are flipping the qubits if there are more than one terms?

Hi Tanuj, may I follow up on this PR please? Thanks!

**Comment by tanujkhattar** [comment]
Hey, apologies for the long delay! Please find my thoughts below:

> Option-3: Y(q[1]) * X(q[0]) + Z(q[0]), i.e only change the order of multiplication, since the order of matrices matters in the Kronecker product, and will give different matrices as output.

The order of matrices also determines what operation you are applying? Using your example of Pauli Strings:
```
ps2 = cirq.X(q[0]) * cirq.Y(q[1])
assert (ps2.matrix([q[1], q[0]]) == ps2.with_qubits(q[1], q[0]).matrix()).all()
```
Note that when you modify the order of qubits, either by `ps.with_qubits` or `ps.matrix([])` the resulting effect is the same . `ps.matrix()` is useful if you not only want to modify the order of qubits but also wish to insert identity operations in between for qubits that don't necessarily exist in your PauliSum. Eg:

```
ps = cirq.X(q[0]) * cirq.Y(q[2])
ps.matrix([q[0], q[1], q[2]]) # This returns the matrix for X(0) * I(1) * Y(2)
```

TL;Dr: If your use case only involves shuffling the order of qubits, then `with_qubits` should suffice. But if you also need the functionality of adding extra identity operations for qubits that don't necessarily have an operation, then `ps.matrix(qubits)` would be needed. 

Since PauliString's already support `.matrix()`, we should go ahead and add a similar functionality for `PauliSum`s as well. I'll reopen this PR. Please add some tests for `.matrix()` that show this equivalence between `with_qubits` for shuffling and adding extra identities for qubits that don't have an operation in the PauliStrings.

**Comment by madcpf** [comment]
Thanks a lot Tanuj for the comments and reopening! I've updated the tests. Could you please take a look? Thanks!

File: cirq-core/cirq/ops/linear_combinations_test.py [hunk 1]
```diff
@@ -1189,6 +1190,40 @@ def test_pauli_sum_formatting():
...
+def test_pauli_sum_matrix():
+    q = cirq.LineQubit.range(3)
+    paulisum1 = cirq.X(q[0]) * cirq.X(q[1]) + 4
+    H1 = np.array(
+        [[4.0, 0.0, 0.0, 1.0], [0.0, 4.0, 1.0, 0.0], [0.0, 1.0, 4.0, 0.0], [1.0, 0.0, 0.0, 4.0]]
+    )
+    assert np.allclose(H1, paulisum1.matrix())
+    assert np.allclose(H1, paulisum1.matrix([q[0], q[1]]))
+    # Expects same matrix when change qubits order.
+    assert np.allclose(H1, paulisum1.matrix([q[1], q[0]]))
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/linear_combinations_test.py:1202]: This is confusion as the same matrix would be expected only in the special case where the pauli sum is symmetric wrt the qubit ordering. 

Instead of adding a special case for this, I'd recommend to only have two cases - 

1. Change qubit ordering only : Here we should also assert that `ps.with_qubits(<new order>).matrix()` is same as `ps.matrix(<new order>)`. 
2. Add new qubits : The new matrix size is larger and hence not equal to `with_qubits`.

**Review by tanujkhattar — CHANGES_REQUESTED** [review]

File: cirq-core/cirq/ops/linear_combinations_test.py (hunk 1 above)
Comment (madcpf) [reply cirq-core/cirq/ops/linear_combinations_test.py]: Thanks Tanuj. I've updated the tests to only include those two cases. But I still don't get why ps.with_qubits(<new order>).matrix() is same as ps.matrix(<new order>). Please see the example below:

In [99]: a = cirq.X(q[0]) * cirq.Y(q[1]) * cirq.Z(q[2])
In [108]: m1 = a.matrix([q[1], q[2], q[0]])
In [109]: m2 = a.with_qubits(q[1], q[2], q[0]).matrix()
In [110]: np.allclose(m1, m2)
Out[110]: False

Comment (tanujkhattar) [reply cirq-core/cirq/ops/linear_combinations_test.py]: That's right, my bad. Please ignore my previous comment. I've also opened an issue to highlight the confusing semantics - #4270. 

I'll approve this PR to unblock you now. Apologies for the confusion which led to the long discussion.

**Review by tanujkhattar — APPROVED** [review]

**Comment by madcpf** [comment]
Thanks Tanuj for your time reviewing and merging this!
